library/std/src/time.rs

```diff
@@ -306,11 +279,23 @@ impl Instant {
     /// sleep(Duration::new(1, 0));
     /// let new_now = Instant::now();
     /// println!("{:?}", new_now.duration_since(now));
+    /// println!("{:?}", now.duration_since(new_now)); // 0ns
     /// ```
     #[must_use]
     #[stable(feature = "time2", since = "1.8.0")]
+    #[rustc_inherit_overflow_checks]
+    #[allow(arithmetic_overflow)]
     pub fn duration_since(&self, earlier: Instant) -> Duration {
-        self.0.checked_sub_instant(&earlier.0).expect("supplied instant is later than self")
+        match self.checked_duration_since(earlier) {
+            Some(d) => d,
+            None => {
+                // In debug builds, trigger a panic on None.
+                // This should optimize completely out in release builds.
+                let _ = usize::MAX + 1;
```

Comment:
As an aside, to me as a casual, this is mystifying. Does this do something different than a `debug_assert!(false)`? I believe I understand that we cannot use `debug_assert!()` in `std` because debug builds of user code still link against a release build of std, but surely this approach isn't fundamentally different? Is this specifically targeted at people building their own std who don't want debug assertions in general but do want this particular panic?

Reply:
That's because std is shipped prebuilt with rustc and without debug asserts enabled. But for overflow checks there's the special mechanism `#[rustc_inherit_overflow_checks]` that does enable overflow checks in std when the user crate has them enabled. So we're abusing that mechanism here to allow conditionally panic depending on user configuration.

The documentation is vague by saying "debug mode" since that covers both overflow checks and debug asserts so we can change it in the future from overflow checks to debug asserts.

Reply:
> But for overflow checks there's the special mechanism `#[rustc_inherit_overflow_checks]` that does enable overflow checks in std when the user crate has them enabled.

Keep in mind this can only work with cross-crate codegen, and I can't find anything in the compiler where the attribute implies `#[inline]` or similar (and there's no checks for whether it's misused, either) - so you likely also need `#[inline]` on this function for it to do anything.

I would suggest adding a test that explicitly sets the overflow checks flags and then making sure the test only passes with `#[inline]`, to confirm this.

Reply:
You probably also need `#[track_caller]` to avoid the panic pointing at the implementation, but that part seems more subjective.

Reply:
I asked here https://github.com/rust-lang/rust/pull/89926#discussion_r733149105 and was informed that putting it on the method itself is sufficient.